### PR TITLE
Add pretty-printing df.toString

### PR DESCRIPTION
**Pull request by bryevdv**
Adds basic pretty-printing method to `DataFrame` e.g.

```
> console.log(new DataFrame({'a': a, 'b': b, 'c': c, 'd': d}).toString({maxColWidth: 70}))
  a          b               d
0.0        0.0 ...        -Inf
1.0        1.1 ...         0.0
2.0        2.2 ... 0.693147181
3.0        3.3 ... 1.098612289
4.0        4.4 ... 1.386294361
5.0        5.5 ... 1.609437912
6.0 6.61234567 ... 1.791759469
```

Supports options:
* `maxColumns`
* `maxColWidth`
* `maxRows`
* `width`

Lots of corner cases and "fix one test break another" but I think this is at a good starting point.

### CODE REVIEW

File: modules/cudf/src/dataframe/print.ts [hunk 1]
```diff
@@ -0,0 +1,150 @@
...
+    this.htrunc      = false;
+    this.vtrunc      = false;
+
+    let tmp = this.preprocess(obj);
+    while (this._totalWidth(tmp) > this.width) {
+      this.htrunc = true;
+      const names = [...tmp.names];
+      const N     = Math.ceil((names.length - 1) / 2);
+      tmp         = tmp.drop([names[N]]);
+    }
```
Comment (bryevdv) [inline modules/cudf/src/dataframe/print.ts:71]: This loop could get in to trouble. I see two options:

* bail if new total width ever equals prev total width
* set a min total width and error earlier

File: modules/cudf/src/data_frame.ts [hunk 2]
```diff
@@ -223,6 +224,16 @@ export class DataFrame<T extends TypeMap = any> {
   /** @ignore */
   asTable() { return new Table({columns: this._accessor.columns}); }
 
+  /**
+   *
+   * @param options
+   * @returns void
+   */
+  toString(options?: DisplayOptions): string {
```
Comment (mzegar) [inline modules/cudf/src/data_frame.ts:232]: Could do

`toString(options: DisplayOptions = {}): string {`

and then avoid the `??`, but it's up to what you prefer.

File: modules/cudf/src/data_frame.ts [hunk 3]
```diff
@@ -223,6 +224,16 @@ export class DataFrame<T extends TypeMap = any> {
   /** @ignore */
   asTable() { return new Table({columns: this._accessor.columns}); }
 
+  /**
+   *
+   * @param options
+   * @returns void
+   */
```
Comment (mzegar) [inline modules/cudf/src/data_frame.ts:231]: Need some `docstrings` about this right?

File: modules/cudf/src/dataframe/print.ts [hunk 4]
```diff
@@ -0,0 +1,150 @@
...
+
+import {DataFrame} from '../data_frame';
+import {Series} from '../series';
+import {Int32, Utf8String} from '../types/dtypes';
+import {TypeMap} from '../types/mappings';
+
+export type DisplayOptions = {
+  /**
+   * The maximum number of columns to display before truncation. Truncation may also occur
+   * if the total printed width in characters exceeds `widht
```
Comment (mzegar) [inline modules/cudf/src/dataframe/print.ts:23]:
```suggestion
   * if the total printed width in characters exceeds `width`
```

File: modules/cudf/src/dataframe/print.ts [hunk 5]
```diff
@@ -0,0 +1,150 @@
...
+   *
+   */
+  maxColumns?: number,
+
+  /**
+   * The maximum width in characters of a column when printing a DataFram. When the column *
```
Comment (mzegar) [inline modules/cudf/src/dataframe/print.ts:29]:
```suggestion
   * The maximum width in characters of a column when printing a DataFrame. When the column *
```

File: modules/cudf/src/dataframe/print.ts [hunk 6]
```diff
@@ -0,0 +1,150 @@
...
+import {DataFrame} from '../data_frame';
+import {Series} from '../series';
+import {Int32, Utf8String} from '../types/dtypes';
+import {TypeMap} from '../types/mappings';
+
+export type DisplayOptions = {
+  /**
+   * The maximum number of columns to display before truncation. Truncation may also occur
+   * if the total printed width in characters exceeds `widht
+   *
```
Comment (mzegar) [inline modules/cudf/src/dataframe/print.ts:24]:
```suggestion
   * (default: 20)
   *
```

File: modules/cudf/src/dataframe/print.ts [hunk 7]
```diff
@@ -0,0 +1,150 @@
...
+export class DataFrameFormatter<T extends TypeMap> {
+  private obj: DataFrame;
+  private maxColumns: number;
+  private maxColWidth: number;
+  private maxRows: number;
+  private width: number;
+  private htrunc: boolean;
+  private vtrunc: boolean;
+
+  constructor(options: DisplayOptions, obj: DataFrame<T>) {
```
Comment (mzegar) [inline modules/cudf/src/dataframe/print.ts:57]: minor nit, maybe `input` or `df` or `dataframe` would be a better name than `obj`? I mention this since the class name is `DataFrameFormatter`

File: modules/cudf/src/dataframe/print.ts (hunk 1 above)
Comment (mzegar) [reply modules/cudf/src/dataframe/print.ts:71]: Yeah setting a min value for `width` seems reasonable to me.

File: modules/cudf/src/dataframe/print.ts [hunk 8]
```diff
@@ -0,0 +1,150 @@
...
+    while (this._totalWidth(tmp) > this.width) {
+      this.htrunc = true;
+      const names = [...tmp.names];
+      const N     = Math.ceil((names.length - 1) / 2);
+      tmp         = tmp.drop([names[N]]);
+    }
+    this.obj = tmp;
+  }
+
+  preprocess(obj: DataFrame<T>) {
```
Comment (mzegar) [inline modules/cudf/src/dataframe/print.ts:75]: If you decide on changing the `obj` param name above.

**Review by mzegar — COMMENTED** [review]
Neat stuff, excited to use this myself.

File: modules/cudf/src/data_frame.ts (hunk 3 above)
Comment (bryevdv) [reply modules/cudf/src/data_frame.ts]: https://github.com/rapidsai/node/pull/297/commits/6853b924285a69689aa481378e551aa295b4a947

File: modules/cudf/src/data_frame.ts (hunk 2 above)
Comment (bryevdv) [reply modules/cudf/src/data_frame.ts]: https://github.com/rapidsai/node/pull/297/commits/6853b924285a69689aa481378e551aa295b4a947

File: modules/cudf/src/dataframe/print.ts (hunk 8 above)
Comment (bryevdv) [reply modules/cudf/src/dataframe/print.ts]: https://github.com/rapidsai/node/pull/297/commits/842a5bd781f3a403a745a759df09936256c70944

File: modules/cudf/src/dataframe/print.ts (hunk 7 above)
Comment (bryevdv) [reply modules/cudf/src/dataframe/print.ts]: following pandas, I went with `frame` 

https://github.com/rapidsai/node/pull/297/commits/842a5bd781f3a403a745a759df09936256c70944

**Review by mzegar — APPROVED** [review]
Looks good on my end, feel free to get some feedback from others as well.

File: modules/cudf/src/dataframe/print.ts (hunk 1 above)
Comment (bryevdv) [reply modules/cudf/src/dataframe/print.ts:72]: I will probably add a min `width` check (5 characters?) but there are lots of interactions so I think I will add the clunky `current = prev` check to. But then there's a question of what to do:

* if things get down to zero column / can't get total width < with

Is that an *error*? Or is that "return an empty string"?

Comment (mzegar) [reply modules/cudf/src/dataframe/print.ts:72]: I'd personally return an empty string, however, is there some convention that the `toString()` methods follow in javascript?

Comment (bryevdv) [reply modules/cudf/src/dataframe/print.ts:72]: I think I'm inclined that way as well. @trxcllnt do you have a preference here?

File: modules/cudf/src/dataframe/print.ts [hunk 9]
```diff
@@ -0,0 +1,151 @@
...
+  constructor(options: DisplayOptions, frame: DataFrame<T>) {
+    this.maxColumns  = options.maxColumns ?? 20;
+    this.maxColWidth = options.maxColWidth ?? 50;
+    this.maxRows     = options.maxRows ?? 60;
+    this.width       = options.width ?? 80;
+    this.htrunc      = false;
+    this.vtrunc      = false;
+
+    let tmp = this._preprocess(frame);
+    while (this._totalWidth(tmp) > this.width) {
```
Comment (trxcllnt) [inline modules/cudf/src/dataframe/print.ts:67]: Does the measured size of the rest of the columns change after dropping a column? Seems like we could measure all the columns once up front, then remove the dropped ones from the corresponding sizes list?

File: modules/cudf/src/dataframe/print.ts (hunk 1 above)
Comment (trxcllnt) [reply modules/cudf/src/dataframe/print.ts:72]: We could truncate the last column values if the user requests a pathologically small max width... Maybe our minimum should be 4 -- 1 character of data + an ellipses?

File: modules/cudf/src/dataframe/print.ts (hunk 9 above)
Comment (bryevdv) [reply modules/cudf/src/dataframe/print.ts:67]: Sure that seems possible :+1:

File: modules/cudf/src/dataframe/print.ts [hunk 10]
```diff
@@ -0,0 +1,151 @@
...
+  private maxRows: number;
+  private width: number;
+  private htrunc: boolean;
+  private vtrunc: boolean;
+
+  constructor(options: DisplayOptions, frame: DataFrame<T>) {
+    this.maxColumns  = options.maxColumns ?? 20;
+    this.maxColWidth = options.maxColWidth ?? 50;
+    this.maxRows     = options.maxRows ?? 60;
+    this.width       = options.width ?? 80;
```
Comment (trxcllnt) [inline modules/cudf/src/dataframe/print.ts:62]: Maybe it's just me, but 80 chars as default max total width seems pretty narrow. Why not default max width to Infinity?

File: modules/cudf/src/dataframe/print.ts (hunk 1 above)
Comment (bryevdv) [reply modules/cudf/src/dataframe/print.ts:72]: That's an interesting idea. That is not what pandas does but I can look into it if

File: modules/cudf/src/dataframe/print.ts (hunk 10 above)
Comment (bryevdv) [reply modules/cudf/src/dataframe/print.ts:62]: I don't have strong opinions here, this is just following what pandas has for a default. I am fine setting the default to unlimited if that's your preferece

File: modules/cudf/src/dataframe/print.ts (hunk 9 above)
Comment (bryevdv) [reply modules/cudf/src/dataframe/print.ts:76]: https://github.com/rapidsai/node/pull/297/commits/510c75317dc52450e67e277ca6d54fdd103f9260

File: modules/cudf/src/dataframe/print.ts (hunk 10 above)
Comment (bryevdv) [reply modules/cudf/src/dataframe/print.ts]: default to unlimited ("terminal") width in https://github.com/rapidsai/node/pull/297/commits/d481fb773ce5a61125be49216ef9b5e921bc23aa 

note that there is still a default max cols though.

File: modules/cudf/src/dataframe/print.ts (hunk 1 above)
Comment (bryevdv) [reply modules/cudf/src/dataframe/print.ts:81]: If our going-in position is default to infinite width, then TBH this seems like a fair amount of complexity for a really rare/unimportant case. I did set a min col width of 4, so there will always be at least some output (even if its just a column of elipses). Any objection if I leave it at that?

Comment (trxcllnt) [reply modules/cudf/src/dataframe/print.ts:81]: If I'm reading this right, we'll drop all the columns if the max width is smaller than the width of the smallest column? I can live with that.
